UI order table model

DavideBrambilla98 opened the pull request:
- Creata la classe OrderTableModel
- Modificata la classe OrderModel:
   aggiungto metodo che restituisce la lista di tutti gli OrderModel dal DB e che li inserisce in una TreeMap
- Creata la classe ProductOrderTableModel

DavideBrambilla98 committed:
Aggiunta classe OrderTableModel

Aggiunta la classe e implementati alcuni metodi base
Piccola modifica di refactoring alla classe OrderModel

DavideBrambilla98 committed:
Update OrderModel.java

DavideBrambilla98 committed:
Merge branch 'main' into ui_OrderTableModel

DavideBrambilla98 committed:
Modifica classe OrderModel e OrderTableModel

- aggiunti i notify per segnalare che il modello è stato cambiato
- sistemato il metodo delete dell' orderRecord, così lo cancella anche nella TreeMap
- aggiunto metodo refresh in OrderTableModel
- aggiunto fireTableDataChange per aggiornare la tabella al variare del modello

MassimilianoGervasoni committed:
aggiornamento Box e BoxModel

aggiunto metodo clone a classe box, aggiunte notify e metodi per ottenere tutti i modelli di box

DavideBrambilla98 committed:
Aggiunta classe ProductOrderTableModel e modifica OrderTableModel

- ProductOrderTableModel: Classe per gestire la tabella dei prodotti relativi ad un
 ordine
- Aggiunto metodo in OrderTableModel che restituisce un orderModel dall'indice di
 una riga della tabella

MassimilianoGervasoni committed:
aggiornamento classe ProductModel

standardizzato la struttura della classe  per renderla coesa con le altre classi del progetto

MassimilianoGervasoni committed:
creazione ProductTablemodel

MassimilianoGervasoni committed:
Update import BoxModel e ProductModel

MassimilianoGervasoni committed:
modifica getProdotto

DavideBrambilla98 committed:
ProductOrderTable update

Aggiunti metodi base per la gestione del modello della tabella dei prodotti relativi ad un ordine

DavideBrambilla98 committed:
Merge branch 'ui-prodotto' into ui_OrderTableModel

DavideBrambilla98 committed:
Update ProductOrderTableModel.java

Completata ProductOrderTableModel

DavideBrambilla98 committed:
Modifica date nel DB per la gestione degli ordini

Tolti i constrains nel DB relativi alle date di emissione e completamento
Modificata classe delle date nelle classi relative agli ordini da Date a LocalDate

JohnnyLAmpAz commented on proj_eclipse/src/main/java/smartmag/models/BoxModel.java:
```diff
@@ -83,14 +86,25 @@ protected BoxRecord getRecord() {
+			throws SQLIntegrityConstraintViolationException {
+		BoxModel bm = getBoxModel(b);
+		if (bm == null)
+			bm = new BoxModel(b);
```
`getBoxModel` non restituisce mai null: se non esiste lo crea lui
```suggestion
		BoxModel bm = getBoxModel(b);
```

JohnnyLAmpAz commented on proj_eclipse/src/main/java/smartmag/models/BoxModel.java:
```diff
@@ -83,14 +86,25 @@ protected BoxRecord getRecord() {
 		return record;
 	}
 
+	public BoxModel CreateBox(Box b)
```
```suggestion
	public static BoxModel createBox(Box b)
```

JohnnyLAmpAz commented on proj_eclipse/src/main/java/smartmag/models/BoxModel.java:
```diff
@@ -83,14 +86,25 @@ protected BoxRecord getRecord() {
 	 *                                                  presente nel db
 	 */
-	public void createBox() throws SQLIntegrityConstraintViolationException {
+	public void create() throws SQLIntegrityConstraintViolationException {
```
privato poiché esiste già createBox publico per creare un Box
```suggestion
	private void create() throws SQLIntegrityConstraintViolationException {
```

JohnnyLAmpAz commented on proj_eclipse/src/main/java/smartmag/models/BoxModel.java:
```diff
@@ -144,6 +161,9 @@ private void cambiaProdotto(Prodotto p) {
 			record.delete();
+			record = null;
+			instances.remove(box.getIndirizzo());
+			notifyChangeListeners(null);
```
Non rimuoverei l'istanza dalla mappa, semplicemente si elimina record (dal db e dell'istanza) e si notifica; isSavedInDb ritornerà _false_
```suggestion
			record = null;
			notifyChangeListeners(null);
```

JohnnyLAmpAz commented on proj_eclipse/src/main/java/smartmag/models/OrderModel.java:
```diff
@@ -75,6 +76,18 @@ public Ordine getOrdine() {
+	public ArrayList<ProdottiordiniRecord> getListaProdottiOrdini() {
+		return (ArrayList<ProdottiordiniRecord>) fetchProductOrderRecordListByOrder(
+				ordine).clone();
+	}
```
Non si dovrebbero restituire publicamente i record del DB. Se l'intento era quello di fornire la lista di prodotti e quantità dell'ordine, ricordiamoci che questo è presente come mappa dentro l'oggetto Ordine ottenibile con *getOrdine().getProdotti()*. Se era per uso interno, c'è direttamente _listaProdottiOrdiniRecord_. Quindi si potrebbe farne a meno.
```suggestion
```

JohnnyLAmpAz commented on proj_eclipse/src/main/java/smartmag/models/OrderModel.java:
```diff
@@ -109,15 +123,19 @@ public void updateOrdine(Ordine o)
 
 		if (p.isValid() && q > 0) {
 			createProdottoOrdineRecord(p, q);
+			notifyChangeListeners(null); // evento per notificare il cambiamento
```
già chiamato da _createProdottoOrdineRecord_
```suggestion
```

JohnnyLAmpAz commented on proj_eclipse/src/main/java/smartmag/models/OrderModel.java:
```diff
@@ -139,6 +157,7 @@ protected void updateStatoOrdine(Ordine o, String newStato)
 
 		o.setStato(StatoOrdine.valueOf(newStato));
 		updateOrdine(o);
+		notifyChangeListeners(null); // evento per notificare il cambiamento
```
giò chiamato da _updateOrdine_
```suggestion
```

JohnnyLAmpAz commented on proj_eclipse/src/main/java/smartmag/models/OrderModel.java:
```diff
@@ -147,11 +166,15 @@ public void setStatoInSvolgimento() {
 	 */
-	public void deleteOrdine() throws ParseException {
+	@SuppressWarnings("unlikely-arg-type")
+	public void deleteRecords() throws ParseException {
```
Più chiaro
```suggestion
	public void deleteOrdine() throws ParseException {
```

JohnnyLAmpAz commented on proj_eclipse/src/main/java/smartmag/models/OrderModel.java:
```diff
@@ -147,11 +166,15 @@ public void setStatoInSvolgimento() {
+	public void deleteRecords() throws ParseException {
 		if (orderIsSavedInDb()) {
 			orderRecord.delete(); // DELETE con UpdatableRecord
+			instances.remove(getOrderModelOf(ordine));
```
di nuovo, ci limitiamo ad eliminare record e notificare; l'istanza del modello rimane
```suggestion
			orderRecord = null;
```

JohnnyLAmpAz commented on proj_eclipse/src/main/java/smartmag/models/OrderModel.java:
```diff
@@ -458,8 +485,44 @@ public static OrderModel create(Ordine o)
+		return treeMapFilter(tm);
+	}
+
+	public static TreeMap<Integer, OrderModel> treeMapFilter(
```
```suggestion
	private static TreeMap<Integer, OrderModel> treeMapFilter(
```

JohnnyLAmpAz commented on proj_eclipse/src/main/java/smartmag/models/ProductModel.java:
Non faccio review in quanto poi si mergia la versione aggiornata da altro branch

JohnnyLAmpAz commented on proj_eclipse/src/main/java/smartmag/models/ui/ProductOrderTableModel.java:
Sei andato a gestire la cosa con i Record... Abbiamo a disposizione i modelli, nello specifico quello dell'ordine che ci restituisce l'oggetto Ordine al cui interno troviamo la mappa della "lista della spesa" che associa prodotti alle quantità. Sto per committare la mia versione

JohnnyLAmpAz reviewed (changes requested):
Vedi commenti. In particolare trovo proprio sbagliato il _ProductOrderTableModel_ perché vai ad usare i Record. Ti committo a breve la mia versione.

JohnnyLAmpAz committed:
ProductOrderTableModel rivisitato utilizzando solo OrderModel

JohnnyLAmpAz commented:
Ho committato [(ae7c7f4)](https://github.com/JohnnyLAmpAz/smartmag/pull/12/commits/ae7c7f4a94523ff823da225bc2d80e187e852013) la mia versione di _ProductOrderTableModel_. Contiene anche una finestra di prova con tabella prodotti relativa ad ordine di id `0` (creane uno nel tuo db locale). Nota come utilizzo solo il modello...

DavideBrambilla98 committed:
Modifiche nella richieste in review alla PR #12

Effettuate modifiche alla classe BoxModel e OrderModel richieste nella review alla PR #12
